lib/rubocop/cop/style/redundant_regexp_escape.rb

```diff
@@ -65,12 +65,18 @@ def allowed_escape?(node, char, within_character_class)
           return true if ALLOWED_ALWAYS_ESCAPES.include?(char) || delimiter?(node, char)
 
           if within_character_class
-            ALLOWED_WITHIN_CHAR_CLASS_METACHAR_ESCAPES.include?(char)
+
+            ALLOWED_WITHIN_CHAR_CLASS_METACHAR_ESCAPES.include?(char) &&
+              !char_class_begins_or_ends_with_hypen?(node, index)
           else
             ALLOWED_OUTSIDE_CHAR_CLASS_METACHAR_ESCAPES.include?(char)
           end
         end
 
+        def char_class_begins_or_ends_with_hypen?(node, index)
```

Comment:
It's not very clear what this function does, so perhaps it'd better to document it a bit.